### DISCUSSION TITLE
API Key for Ollama Adapter?

### DISCUSSION THREAD

**Discussion by enthusiva** [olimorris/codecompanion.nvim#127] [Q&A] [answered]
I am trying to connect to a remotely hosted ollama server. I need to pass in an Bearer Auth token to access the Ollama API. I tried set the env variable 'api_key' as recommended for other adaptors in the READ ME file. But it doesnt seem to pickup the api key from the env variable. Please help me to solve this problem.

**Comment by olimorris** [comment]
I advise [troubleshooting](https://github.com/olimorris/codecompanion.nvim#toolbox-troubleshooting) your config, reading the [ADAPTERS](https://github.com/olimorris/codecompanion.nvim/blob/main/doc/ADAPTERS.md) guide and seeing how I've setup other [adapters](https://github.com/olimorris/codecompanion.nvim/tree/main/lua/codecompanion/adapters) in the plugin.

As a word of warning, I always recommend putting API keys into something more secure than your shell. I use 1Password and document this in the README>

**Reply by olimorris** [reply]
@enthusiva any luck?

**Reply by enthusiva** [reply]
@olimorris , thanks for your guidance. I had look at the adatptor code and able to get it working. But I am not sure, this is the right way to get this done :)

```lua
      adapters = {
        llama3 = function()
          local remote_url = 'my_remote_url_here'
          local api_key = os.getenv 'OLLAMA_API_KEY'

          local function get_models()
            local curl = require 'plenary.curl'
            local response = curl.get(remote_url .. '/v1/models', {
              sync = true,
              headers = {
                ['content-type'] = 'application/json',
                ['Authorization'] = 'Bearer ' .. api_key,
              },
            })
            if not response then
              return {}
            end

            local ok, json = pcall(vim.json.decode, response.body)
            if not ok then
              return {}
            end

            local models = {}
            for _, model in ipairs(json.data) do
              table.insert(models, model.id)
            end

            return models
          end
          return require('codecompanion.adapters').extend('ollama', {
            name = 'llama3', -- Ensure the model is differentiated from Ollama
            env = {
              url = remote_url,
              api_key = api_key,
            },
            headers = {
              ['content-type'] = 'application/json',
              ['Authorization'] = 'Bearer ${api_key}',
            },
            schema = {
              model = {
                default = 'llama3.1:70b',
                choices = get_models(),
              },
              num_ctx = {
                default = 16384,
              },
              num_predict = {
                default = -1,
              },
            },
          })
        end,
      },

```

**Reply by olimorris** [reply, marked as answer]
Try this:

```lua
require("codecompanion").setup({
  adapters = {
    ollama_remote = function()
      return require("codecompanion.adapters").extend("ollama", {
        name = "ollama_remote",
        env = {
          url = "https://my_ollama_url",
          api_key = "OLLAMA_API_KEY",
        },
        headers = {
          ["Content-Type"] = "application/json",
          ["Authorization"] = "Bearer ${api_key}",
        },
        parameters = {
          sync = true,
        },
        schema = {
          model = {
            default = "llama3.1:70b",
          },
          num_ctx = {
            default = 16384,
          },
          num_predict = {
            default = -1,
          },
        },
      })
    end,
  },
  strategies = {
    chat = {
      adapter = "ollama_remote",
    },
    inline = {
      adapter = "ollama_remote",
    },
    agent = {
      adapter = "ollama_remote",
    },
  },
})
```

The `choices` method should automatically use the url in your env table.

**Reply by enthusiva** [reply]
@olimorris I tried this setup, the problem is the get_models function  in the ollama adapter is not sending the api_key header  while checking if ollama is running or not. I get this error 'Ollama is not running or cannot be found on localhost:11434' as it gets 401 response in return.

**Reply by olimorris** [reply]
Great spot! Yes of course it needs to authenticate in order to pull the models down. Leave that with me. That's a bug in how I've setup the adapter.

**Reply by olimorris** [reply]
@enthusiva this should now be fixed in the latest commit and the code I provided above should work.

**Reply by enthusiva** [reply]
I have updated the plugin to latest commit but still getting the 'Could not parse the response from ...' message when I try to view the list of models from ollama.

**Reply by olimorris** [reply]
I've just pushed an update which outputs the response. Let me know what it says. I can't fathom why it's not working

**Reply by enthusiva** [reply]
I received the 'Invalid API Key' error response. The issue is self.args.env.api_key is not replaced with the actual key from the environment, instead it holds 'OLLAMA_API_KEY'.

**Reply by olimorris** [reply]
You need to put your own key in. That key doesn't exist in the environment so it's treating it as a plain string.

**Reply by enthusiva** [reply]
I am able to chat with the ollama remote server without any issues. this error occurs only when I try to pull the models list from the remote server. I also verified that the OLLAMA_API_KEY is set properly in my shell environment.

**Reply by olimorris** [reply]
Argggh. Apologies...should finally be fixed now. Thanks for bearing with me.

**Reply by enthusiva** [reply]
Yeah. It works now. Thank you very much.!!

**Reply by olimorris** [reply]
Wooop. We got there in the end. Sorry for being so slow. Long day 😆

**Reply by enthusiva** [reply]
Just a quick suggestion after looking at your fix. Can you please use the headers from the schema rather than using hardcoded bearer auth inside the get_models function? I use Bearer auth but other may use a different header for sending the api key. so It would be flexible for everyone if you could use the headers provided from the extended schema for ollama?

**Reply by olimorris** [reply]
Can you share an example of what that looks like?

I'm moving to an "unofficial" policy of implementing fixes when they get raised. So will likely implement that if someone faces that issue.

**Reply by enthusiva** [reply]
Since ollama server does not have native auth support, everyone out there has to run it behind a proxy for remote access, where the auth header may be different. one may use 'Bearer' and other may use a different key like 'X-API-Key'. It is upto you whether to implement it in this way or not. :) Really appreciate your great work. Sorry if I had asked for too much. Also please bear with my bad english.
